[BB-112] Fix the order randomization behaviour of Randomized Content Block

lgp171188 opened the pull request:
This PR (supersedes PRs #23265, #20723, #18675 which were merged and then reverted due to issues) contains changes to the Randomized Content Block XBlock. The XBlock has unpredictable randomization of the order of the selected child blocks due to the usage of sets, which are unordered, for storing the selected blocks. This becomes apparent when all the available child blocks in a library are chosen for a Randomized Content Block to randomize just the order of the child blocks. 

This PR modifies the XBlock to store the selected child blocks in a list after randomly shuffling them.

The `ContentLibraryOrderTransformer`, a new block structure transformer added in this PR, uses the order of the selections made in the existing `ContentLibraryTransformer` to re-order the blocks. However, as other transformers affecting the block structure (like those used for content type gating, visibility etc.) can run after the `ContentLibraryTransformer` and before the `ContentLibraryOrderTransformer` and modify the block structure to make the stored selections invalid, the `ContentLibraryOrderTransformer` transforms the order only when the stored selections and the current children of the randomized content block match. Otherwise, it just logs an informational message when this happens.

**JIRA tickets**:  None

**Dependencies**: None

**Sandbox URL**:  TBD

**Merge deadline**: None

**Testing instructions**:
1. Login to the CMS as a user who can create and edit courses.
1. Create a library with a few child components.
1. Add a unit to a course.
1. Add a Randomized Content Block to the unit.
1. Select the library created above as the source for this XBlock.
1. Select all types of children to be shown.
1. Specify the total number of components present in the library as the number of components to display.
1. Save the unit and publish it.
1. Register a few students to this course.
1. Navigate to the unit containing this Randomized Content Block.
1. Verify that the order of the content blocks are randomized for every student.
1. Also verify that the same order of children blocks is returned by the `course_blocks` API for that randomized content block.
1. Add some type of content gating/visibility changes to make the randomized content block hidden from the test user.
1. Verify that there are no exceptions thrown when the LMS returns the block structure for that user - for example, when opening the course outline page. Also verify that the expected hidden blocks stay hidden.

**Reviewers**
- [ ] @kaizoku 
- [ ] edX reviewer[s] TBD (@ormsbee maybe?)

edx-webhook commented:
Thanks for the pull request, @lgp171188! I've created [OSPR-4773](https://openedx.atlassian.net/browse/OSPR-4773) to keep track of it in JIRA, where we prioritize reviews.  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

Please let us know once your PR is ready for our review and all tests are green.

natabene commented:
@lgp171188 Thank you for your contribution. Please let me know once it is ready for our review.

kaizoku commented on lms/djangoapps/course_blocks/transformers/library_content.py:
```diff
@@ -177,3 +188,68 @@ def publish_event(event_name, result, **kwargs):
+                if current_children_blocks != current_selected_blocks:
+                    logger.info(
+                        u'Mismatch between the children of %s in the stored state and the actual children for user %s. '
+                        'So not transforming the order.',
```
nit: Does "Abortin order randomization" make sense here? The new message is a bit unclear to me.
Or perhaps "Continuing without order randomization" ?

Also, I see the re-shuffling has been removed in this section. I tested a couple scenarios, and if I add new problems to the content library, a user who has already viewed the course will not see a change in problems, *unless* the problem count is changed in the randomized content block. This doesn't seem like a problem, but does it change the behavior from before?

lgp171188 replied:
@kaizoku,

> nit: Does "Abortin order randomization" make sense here? The new message is a bit unclear to me.
Or perhaps "Continuing without order randomization" ?

:+1:. I will use `order transformation` as this transform re-orders the blocks to match the randomized order in selection made in the existing `ContentLibraryTransformer`.

`ContentLibraryTransformer` makes the selection,  shuffles the children blocks and then stores them the student module. `ContentLibraryOrderTransformer` retrieves the stored selections and re-orders the children blocks to match the order in the retrieved data.

> Also, I see the re-shuffling has been removed in this section. I tested a couple scenarios, and if I add new problems to the content library, a user who has already viewed the course will not see a change in problems, unless the problem count is changed in the randomized content block. This doesn't seem like a problem, but does it change the behavior from before?

In the existing implementation, the existing selection is changed only when the `count` is changed or if any of the selected blocks are no longer valid (i.e., they have been removed from the library). This PR preserves the same behaviour and does the selection and shuffling only in these 2 cases.

The goal of the changes in this PR is to shuffle the order of the selected blocks so that the order is not similar due to set ordering. Does this make sense to you?

kaizoku replied:
Yes, that makes sense @lgp171188. I think I was just seeing a bit of code removed from this section compared to the previous PR and wanted to make sure it didn't change the behavior.

Thanks for updating the logging string as well.

bradenmacdonald commented on common/lib/xmodule/xmodule/library_content_module.py:
```diff
@@ -158,37 +158,40 @@ def make_selection(cls, selected, children, max_count, mode):
+            selected = selected_keys
 
         return {
             'selected': selected,
```
Nit/Q: Can we make this `'selected'` return value a shuffled list, so that the

```python
            selected = list(block_keys['selected'])
            random.shuffle(selected)
```
code is not duplicated in both `LibraryContentModule.selected_children` and `ContentLibraryTransformer.transform_block_filters` ?

lgp171188 replied:
This is fixed now.

bradenmacdonald commented on lms/djangoapps/course_blocks/transformers/library_content.py:
```diff
@@ -177,3 +188,68 @@ def publish_event(event_name, result, **kwargs):
+    requires the selections for the randomized content block to be already
+    made either by the ContentLibraryTransformer or the XBlock.
+
+    Staff users are *not* exempted from library content pathways/
```
Nit: I'm not sure what "pathways" means here nor why it ends with a `/`

lgp171188 replied:
@bradenmacdonald, the traling `/` is a typo and I will fix it. I have just copied the text from the `ContentLibraryTransformer` [docstring](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/course_blocks/transformers/library_content.py#L29) since it also applies to the related `ContentLibraryOrderTransformer.

lgp171188 replied:
I have fixed the typo.

bradenmacdonald commented on lms/djangoapps/course_blocks/transformers/tests/test_library_content.py:
```diff
@@ -167,3 +168,159 @@ def test_content_library(self):
+    @mock.patch('lms.djangoapps.course_blocks.transformers.library_content.get_student_module_as_dict')
+    def test_content_library_randomize_selected_blocks_mismatch(self, mocked):
+        # The block structure will contain all the blocks in the course. So the selections
+        # returned from the database will trigger a mismatch
```
Could you please move this comment into the docstring and expand the docstring add more explanation of this test? Without the context that you have in the PR description, this test is pretty confusing.

lgp171188 commented:
@bradenmacdonald, 

> I have read through the code and it looks good - just a few minor requests. I will test+finish review tomorrow. Would you be able to rebase it and provide a link to the updated sandbox (with some useful test course) so I can test it on the sandbox?

Thanks for the review! 🙂 

> BTW Is there a way to see what exactly you changed since the last PR? In the future, I'd prefer if you can keep the changes as a separate commit (during code review) so it's easy to see what the new fix is compared to the previous PR; then we can squash before merge.

I managed to cleanup the PR to break it down into separate commits corresponding to the previously reverted PRs and added the changes made for this PR on top of those changes. However, since I had to rewrite history to do this, is it okay to force-push the PR branch to replace it? The existing review comments may become difficult to track once that is done.

> What determines the order in which the transformers run? From the code I would have expected ContentLibraryOrderTransformer to run immediately after ContentLibraryTransformer.

The filtering transformers are all clubbed into a single big transformer (see [here](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/content/block_structure/transformers.py#L134) and [here](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/content/block_structure/transformers.py#L111-L114)) and run in a single tree traversal and then the block structure transformers (aka the non-filtering transformers) are all run one by one in the order they are defined.

So it is likely (and appears to be the case) that there are additional filtering transformers that run after the `LibraryContentTransformer` and remove the children blocks selected by the `LibraryContentTransformer`.

Below is the explanation I had given for the `KeyError` issue seen with the previous iterations of this PR and the fix in a direct email thread about this PR.

>When a randomized content block is gated from the user using various features like visibility, content gating etc., the corresponding transformers run after the 'ContentLibraryTransformer' and before the 'ContentLibraryOrderTransformer'. Due to this, the selected blocks stored in the 'courseware_studentmodule' table are no longer present in the block structure. So when the 'ContentLibraryOrderTransformer' tries to get those blocks, they are not present in the block structure any more and hence throw a 'KeyError'.
>
> To fix this issue, I have updated the 'transform()'  method of the 'ContentLibraryOrderTransformer' to do its order transformation only when the stored selections match the current children of the randomized content block. If there is a mismatch, which indicates that the children blocks may not be available in the block structure any more, it just logs a message and does nothing.

bradenmacdonald commented:
@lgp171188 Thanks, those last changes look good except for one issue: https://github.com/open-craft/edx-platform/commit/edbdf1b0a156494fe6f440d3b644c60767f53f70#r41255356

Please go ahead and force-push to this PR. GitHub should preserve all the comments.

Before final approval I will also need to test this on a sandbox - can you please update the https://pr24343.sandbox.opencraft.hosting/ server to the latest version and provide me a link to test it?

We'll also need to squash down to one commit before merging. You can do that now too if you want, but first force-push your existing branch to the PR, so that when you squash and force push a second time, GitHub will show a diff of any changes (such as fixing the issue I mentioned above).

lgp171188 commented:
@bradenmacdonald,

> @lgp171188 Thanks, those last changes look good except for one issue: open-craft@edbdf1b#r41255356

I saw your comment already and am fixing it now. 👍🏼 

>Please go ahead and force-push to this PR. GitHub should preserve all the comments.
> We'll also need to squash down to one commit before merging. You can do that now too if you want, but first force-push your existing branch to the PR, so that when you squash and force push a second time, GitHub will show a diff of any changes (such as fixing the issue I mentioned above).

I will force push the changes in that separate branch to this branch now and will squash and force push a second time after you have finished your testing and there are no more changes to make.

> Before final approval I will also need to test this on a sandbox - can you please update the https://pr24343.sandbox.opencraft.hosting/ server to the latest version and provide me a link to test it?

I have spawned a new appserver on the sandbox to get it ready for testing. Will let you know once it is ready for testing.

lgp171188 commented:
@bradenmacdonald, the sandbox, https://pr24343.sandbox.opencraft.hosting/, is up and has the latest changes from this PR deployed.

lgp171188 commented:
> +1 Thanks @lgp171188. Now that I've reviewed this, please squash it down to a single commit (include useful context from the PR description in the commit message). 

Thanks @bradenmacdonald for the quick review cycle. I have squashed the changes into a single commit with a very detailed description in the commit message. Please let me know if you want the message to altered.

>We'll merge it on Monday; given that this has been reverted several times already I don't want to merge on a Friday :p

👍🏼

kdmccormick commented:
Per a conversation with @ormsbee , I'm merging this and will be monitoring it as it rolls out today. Fingers crossed!

edx-webhook commented:
@lgp171188 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/24343&entry.1484655318=https://github.com/lgp171188&entry.752974735=2020-06-29+07:54&entry.1917517419=2020-08-11+15:46&entry.2133058324=Yes) so we can improve your experience in the future.

bradenmacdonald commented:
Thanks @kdmccormick! Fingers crossed.

kdmccormick commented:
No noticeable issues yesterday. I think this is in the clear :tada:

lgp171188 commented:
Thanks for the good news @kdmccormick 🙂 Please let us know if this causes issues in the future.

kdmccormick commented on lms/djangoapps/course_blocks/transformers/library_content.py:
```diff
@@ -177,3 +180,66 @@ def publish_event(event_name, result, **kwargs):
+                # has modified those blocks (for example, content gating may have affected this). So do not
+                # transform the order in that case.
+                if current_children_blocks != current_selected_blocks:
+                    logger.info(
+                        u'Mismatch between the children of %s in the stored state and the actual children for user %s. '
+                        'Continuing without order transformation.',
+                        str(block_key),
+                        usage_info.user.username
+                    )
```
Hey @lgp171188. This statement is blowing up our logs -- we're seeing 250k logs/hour from it.

Is it alerting us of something important that we should fix? Or can we remove it or turn it to `debug` level, which will exclude it from our production logs?

lgp171188 replied:
@kdmccormick, this PR was merged after 3 reverts due to issues in this transformer and that is why this logging statement was added. But now that the PR is merged and the changes have worked okay in production for a while now, it should be okay to remove it or turn it to the `debug` level. 👍🏼 Sorry for causing this issue!

kdmccormick replied:
No sweat! Can you give https://github.com/edx/edx-platform/pull/25065 a look? It sets the logging level to debug.
